### DISCUSSION TITLE
Time-warping stability

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#777] [Q&A]
[Paper](https://arxiv.org/pdf/1304.6763.pdf) (pg 2) reads

<img src="https://user-images.githubusercontent.com/16495490/131631407-c3a2d0df-196e-40f5-a297-3224c5883d3a.png" width="450">

What is the `|tau'| < 1` about? Seems excessively permissive:

![image](https://user-images.githubusercontent.com/16495490/131632913-00404a64-f515-420b-b6f5-3580555ce85c.png)

That's a pure sine of `f` = 2048Hz with 8192 samples, deformed by [this func](https://www.desmos.com/calculator/mw23ztqedl).

<details>
  <summary><b>code</b></summary>

```python
import numpy as np
from kymatio.numpy import Scattering1D
from kymatio.visuals import imshow

def tau(t, K=.08):
    return np.cos(2*np.pi * t**2) * K

def adtau(t, K=.08):
    return np.abs(np.sin(2*np.pi * t**2) * t * 4*np.pi*K)

#%%###########################################################################
N, f = 8192, 2048
K0, K1 = .03, .09
t = np.linspace(0, 1, N, 0)
x0 = np.cos(2*np.pi * f * (t - tau(t, K=K0)))
x1 = np.cos(2*np.pi * f * (t - tau(t, K=K1)))

#%%###########################################################################
J, Q = 10, 16
ts = Scattering1D(shape=N, J=J, Q=Q, average=False, oversampling=999,
                  out_type='list')
meta = ts.meta()

#%%###########################################################################
Scx0 = ts(x0)
Scx1 = ts(x1)
S0 = np.vstack([c['coef'] for c in Scx0])[meta['order'] == 1]
S1 = np.vstack([c['coef'] for c in Scx1])[meta['order'] == 1]

#%%###########################################################################
mx0, mx1 = adtau(t, K0).max(), adtau(t, K1).max()
imshow(S0, abs=1, title="max(|tau'|) = %.2f" % mx0)
imshow(S1, abs=1, title="max(|tau'|) = %.2f" % mx1)
```
</details>

The way I see it, it bounds phase such that `f_tau = 0 to 2*f` (`f*t --> f*(t - tau)`, derivative `f --> f*(1 +- 1)`). There's no condition on higher-order derivatives, only that `tau << T`, which is satisfied here. Thus `x_tau` can assume nearly any FM behavior, e.g. down chirp, so if distance to such deformation is claimed possible to be small, it's same as claiming scattering can't distinguish any two unimodal signals of same amplitude.

The explicit statement is, there's a `C > 0` such that 

<img src="https://user-images.githubusercontent.com/16495490/131635126-2e899401-24cf-4016-914f-89468466c82b.png" width="300">

but this holds even with zero coefficient overlap so long as `x` has finite norm, which seems an ill indicator of "stability". It's also strange for `max(|tau'|)` to not depend on `T`.

**Comment by lostanlen** [comment]
`|tau'(t)| < 1` is more than an hypothesis: it supports the definition of `x_tau` as a warped version of `x`.
Take the opposite case where `tau'` is allowed to be above 1. Then there is no guarantee that `t - tau(t)` is an increasing function anymore. In other words, time is not only warped but locally reversed.
The theorem you're citing only applies to warped signal, not reversed ones; hence tau' should obviously be smaller than 1.

Now, in practice, we need more than this. At the limit `tau'(t) --> 1`, the warped time `t - tau(t)` becomes extremely slow, to the point of becoming motionless. If `x` is a sine wave, this induces a frequency modulation, as you showed in the example above. (this is comparable to a Doppler effect)
To ensure stability, we need the amount of relative frequency modulation (Delta_f / f) to be smaller than the relative bandwidth of the wavelet (Delta_omega / omega), that is, the inverse of the quality factor Q. (Q = omega / Delta_omega)

So the condition is `tau'(t) << 1/Q` which itself implies `tau'(t) < 1` because Q is necessarily larger than 1.
It seems like you've chosen a big Q in your example so make sure to reduce your sup tau' accordingly

(EDIT: should have been `|tau'|(t) << 2**(-1/Q)`, not `|tau'|(t) << 1/Q`)

I hope this helps!

**Reply by OverLordGoldDragon** [reply]
Helpful, thanks. Some followups:

  1. _"freq mod < rel bandwidth"_ equivalently means "warped and original scalograms overlap", correct?
  2. Clarifying for readers, Kymatio's Q `!=` Q in literature; Kymatio's Q sets the number of wavelets per octave, and affects actual Q indirectly. The relation is `Q_lit = (1 - .5**(1/Q_kym)) / (1 + .5**(1/Q_kym)) * 1/sqrt(2*log(1/r_psi))`, where `r_psi` (default `sqrt(.5)`) controls `sigma_max`. 
     - _"Q is necessarily larger than 1"_ refers to `Q_lit`; if it were false, e.g. `omega=.5` wavelet's bandwidth would exceed .5 and tile the entire frequency plane. _This is possible_, and occurs in extreme time localization applications, but never in scattering. Strictly speaking then we have `tau' < min(1/Q, 1)`.
   3. The lefthand of relation with `C`, assuming tight frame, peaks at `2*||x||`, i.e. zero overlap - correct? (each `Phi` by itself conserves input's energy) The measure then remains blind to extent of deformation, i.e. simple frequency transposition is same as an arbitrary frequency modulation, so it doesn't fully quantify "stability"
   4. _"big Q ... reduce sup tau'"_ - the alternative is to also bump up `r_psi`, but that's undesired per raising redundancy.

**Reply by OverLordGoldDragon** [reply]
5. Is there a relationship for `T` and `tau'`? I confirmed larger `T` increases warp invariance for small warps, consistent with Section 2A showing an equivalence of time and frequency averaging. Paper cites `|tau| << T`, but we can have that with large `|tau'|`. So a relation between scalogram freq and time averaging, rather than spectrogram freq and scalogram time.

**Reply by lostanlen** [reply]
1. yes
2. here i'm referring of Q as the ratio of the center frequency to the bandwidth, not the number of filters per octave (which may be bigger than such ratio)
3. yes, by energy conservation and triangular inequality, one has the following upper bound
`|| Phi(x) - Phi(x_tau) || <= || Phi(x) || + || Phi(x_tau) || <= || x || + || x_tau ||`
Then you need to look at the energy of x_tau. I'm not sure that the definition above satisfies `|| x || = || x_tau ||` for any tau.
For that, i would expect a definition of the like of `x_tau(t) = (1-tau'(t)) x(t-tau(t))`
So that a change of variable `u = t-tau(t)` leads to `du = (1-tau'(t)) dt`
I could be wrong, it's worth checking.
4. Right. It's also that we don't want to change the definition of the quality factor (via changing r_psi) while changing the quality factor itself (via changing Q). r_psi should be kept constant
5. yes, even if `tau'(t)<<1`, T also plays a role if `|t - tau(t)|` takes large values for some t. See Eq (2.48) in Group Invariant Scattering. There is a `2^-J ||x||_infty` term in addition to a gradient term and a Hessian term.

**Reply by OverLordGoldDragon** [reply]
2) Right, `Q_lit`
3) Indeed under 2.61 reads `||Phi(x_tau)|| <= ||x|| * (1 - max(||tau'||))^(-d)`
4) `sigma_max` is sometimes a needed degree of freedom, so I'd say "depends". The formula for `Q_lit` remains the same and non-linearly relates to `Q_kym` anyway so think not much changes on definitions.
5) By standard scattering we have `log2_T == J` and thus `2^-J` contracts the upper bound on Euclidean distance - makes sense. I notice the theorem uses `|tau'| < 1/2` and, under 2.61, states it can be replaced by `< 1` but that makes `C` depend on `|tau'|`, which seems circular? So which is more appropriate?

**Comment by OverLordGoldDragon** [comment]
Dug some - I present my thought process and conclusion.

I can't see how Lipschitz continuity adds anything to contraction; if distance norm is bounded by input norm, there always "exists a constant". My experiments show, if anything, distance grows fastest when displacement is small - which makes sense, since that's when partials begin to un-overlap and most of a wavelet's energy is concentrated near its center. Then the only thing `|tau'| < 1/Q` seems to guarantee is that the distance will be small - which is invariance rather than stability. 

The only point of a `C` is if `C<1` when `K(tau) = 1` and `||x|| + ||x_tau|| = 1`. It seems, instead, the point is: coefficient distance is bound to _size of deformation_, not _shape of signal_. That is, we can't find `x` whose coefficient distance with `x_deformed` exceeds said `C < 1` bound; this isn't true for STFT, where we can beat any `C` by frequency-shifting `x`. Question is, what does `C` depend on, exactly? The most I found is, under [Eq 24](https://arxiv.org/pdf/1304.6763.pdf):

<img src="https://user-images.githubusercontent.com/16495490/133692700-f1c63d31-265b-460c-99ad-2394d82237ad.png" width="400">

so `Q` plays a role, but in `C` (as `Q` rather than `1/Q`) instead of in `|tau'|` - and the original `|tau'| < 1` seems to hold. Regardless `C` must be as aforementioned, so the question stands - what exactly determines `C` and is it ever possible for the right bound to exceed `||x||`?

The answer is no, and `C` is determined by the filterbank. Following [Theorem 2.12](https://arxiv.org/pdf/1101.2286.pdf), what we have is: 

 ```
||Sx - Sx_tau|| <= const(filterbank) * K(tau) * ||x|| <= ||x|| + ||x_tau||
```
So the true statement of "stability" is: difference in representations is bounded _only_ by size of deformation, _for all signals_. No matter what `x` we pick, its coefficient distance with `x_tau` is limited by `tau` - and this limit can lie well below `||x|| + ||x_tau||`. This is untrue of STFT, where we can always drive distance to `||x|| + ||x_tau||` by frequency-shifting `x` higher.

Meanwhile "invariance" says, representations are nearly same for all deformation sizes under the scale of invariance, for all signals. This can be achieved in STFT.

My experiments confirm above stability statement.

## Math details

The theorem of interest is 

<img src="https://user-images.githubusercontent.com/16495490/133695090-18e76263-cebd-408a-ab78-3e4c07526f58.png" width="550">

Proof begins with the triangular inequality, `||Sx - Sy|| <= ||Sx - Sz|| + ||Sy - Sz||`, warranted by contraction/non-expansiveness:

<img src="https://user-images.githubusercontent.com/16495490/133693386-ed36c516-81cd-4d43-b842-a9a98eb87aae.png" width="550">

The first RHS term

<img src="https://user-images.githubusercontent.com/16495490/133693506-701a01f5-e2c0-4ce7-91f5-0b599731e15c.png" width="550">

evaluates per an earlier derivation; `A_J` is lowpass filtering:

<img src="https://user-images.githubusercontent.com/16495490/133693581-520bb750-3f8b-4d5a-9aae-abec488f6465.png" width="550">

Next,

<img src="https://user-images.githubusercontent.com/16495490/133693701-9962dde0-e847-4399-898c-ad661a721539.png" width="550">

Again, per non-expansiveness, the RHS is bounded by `||LA - A|| ||x||`: **lowpass filters**, and input _norm_. What remains is the second term:

<img src="https://user-images.githubusercontent.com/16495490/133693945-7224629e-b5d8-4fc8-b45c-8b62d56efc69.png" width="550">

Hence we must show `||W_J, L_J||`'s bound; this is done in Appendix E, using the following lemma:

<img src="https://user-images.githubusercontent.com/16495490/133694431-65b6a0e3-dbeb-4922-8f54-df6c5472f75f.png" width="550">

derivation hence proceeds with general `h` kernels, whose described decay properties are what wavelets satisfy. `C` is used throughout the proof - but the key is:

<img src="https://user-images.githubusercontent.com/16495490/133694947-e98b4d33-748a-45f7-af6c-269448b48fcb.png" width="550">

Hence, the other part of the constant is determined by **wavelet filters**. The entire constant is thus determined entirely by the filterbank, independent of characteristics of `x` besides its norm.

**Reply by lostanlen** [reply]
i forgot what the question was?

**Reply by OverLordGoldDragon** [reply]
Our discussion on "stability" vs "invariance" and how exactly they differ despite both using Euclidean distance; we agreed former is "grows with size of deformation" but that left much room for "what if"s, also the `|tau'| < 1/Q` condition seemed off - so I took another look and the picture's clearer now.

**Reply by OverLordGoldDragon** [reply]
Summarizing, 1) `C` is determined entirely by the filterbank and isn't arbitrarily large; 2) `|tau'| < 1/Q` isn't correct, or it's for something else, but `Q` is involved, in `C`;  3) theorem's meaning is disambiguated by comparing against STFT and proving the upper bound is independent of `x` besides norm.

**Reply by OverLordGoldDragon** [reply]
@lostanlen Is `1/Q` for linearizing additive warps? The stability theorem derives for all `|tau'| < 1`, but I don't see a description on region of linearity, and `Q` seems absent from the paper. Deep SS mentions `sup|tau| << T`, and I can  see `1/Q` being its log-frequency equivalent, but I've not found an explanation. To note, pure tone frequency shifts, and impulse time shifts, are linearized without limits.

**Reply by lostanlen** [reply]
honestly i don't know how these theorems interact with the choice of `Q`. I imagine that increasing `Q` increases discriminability and reduce stability? i haven't done the math. There's probably a corollary of `2.12` to be written with arbitrary `Q>=1`.

>  C is determined entirely by the filterbank and isn't arbitrarily large

yes

> |tau'| < 1/Q isn't correct

what do you mean, isn't correct? this is the assumption, not the result. how can it be not correct? is the assumption too weak? what would you assume instead?

>  Is 1/Q for linearizing additive warps? 

i don't understand. What's an additive warp?
`1/Q` is the log-frequential bandwidth (in `octaves^(-1)`) of each wavelet.

> Q seems absent from the paper.

Yes, the whole GIS paper takes `Q=1`. It's complicated enough like that! haha

> DSS mentions sup|tau| << T

Right. sup |tau|  is the maximum amount of time shift, so it makes sense that we should compare it to `T` (our chosen amount of time shift invariance).

> but I've not found an explanation

our deformation model is `t --> a * t + b` with `|b|` less than `T` and `|a|` somewhere being `2**(-1/Q)` and `2**(1/Q)` (i'm guessing the bounds here, but i think it's about right).

> To note, pure tone frequency shifts, and impulse time shifts, are linearized without limits.

Can you explain?

**Reply by OverLordGoldDragon** [reply]
> what do you mean, isn't correct? this is the assumption

The assumption is `|tau'| < 1` in the theorem. Granted, if there's a silent `Q` that defaulted to `1`, then it's unclear. Where is it shown to be the assumption? `1/Q` would make the "stability" much weaker, and I confirmed it to work for `>>1/Q` on a wide variety of synthetic signals - but not fully thoroughly.

> What's an additive warp?

`t - tau` as opposed to `tau * t`

> Can you explain?

Misspoke there; the Euclidean distance will grow linearly within `T` and `1/Q`, and even `2*T` and `2/Q` approximately - simply distance of Gaussian with its shifted self. 

The point is, and is at root of my challenge to `1/Q`, that the distance will grow with size of deformation, below or over `1/Q`. CQT enforces a "uniformity" along log-frequency, and temporal lowpassing along time - that is, a _common minimal displacement support_.

  - In STFT, shifting in frequency from higher frequency will shift us more; no common frequency displacement (equivariance)
  - In CWT (CQT), shifting in time from higher frequency will shift us more; no common temporal displacement (equivariance)
  - Scattering inherits CQT for frequency - and for time, at each scale, the "pseudo impulse response" has support of `T + psi_lambda`, so no matter where we move, we guarantee a displacement field that grows at most against `T`.

Gaussian lowpassing a Gaussian (wavelet modulus envelope) doesn't change how the distance of resulting waveform grows relative to itself shifted, only shifts the bounds (takes longer to reach max distance). Hence scattering introduces a _guarantee_, no matter at which frequency we are, the maximum distance will be attained only after a sufficiently large displacement in time and/or frequency - and that exactly seems to be the point of the stability theorem (`x`-independence, "there exists a `C`").

Hence the only role of `1/Q` seems to be to guarantee linearity - i.e. distance grows exactly proportionally with size of deformation.

**Reply by lostanlen** [reply]
> The assumption is |tau'| < 1 in the theorem. 

Yes

> "Granted, if there's a silent Q that defaulted to 1, then it's unclear"

Yes, that's the case, but what makes it unclear?


> Where is it shown to be the assumption? 

Right after the introduction, on page 4:

<img width="894" alt="Screenshot 2021-09-19 at 22 05 01" src="https://user-images.githubusercontent.com/3943142/133941465-bce4db10-33d0-4bd1-a277-f7cf9e750cc3.png">

(FYI: `a = 2^(1/Q)`). 


> 1/Q would make the "stability" much weaker, and I confirmed it to work for >>1/Q on a wide variety of synthetic signals - but not fully thoroughly.

I don't understand what you're talking about. The theorem applies to all signals, not just synthetic ones; so it doesn't compare with your experiments. Besides, `Q` is a parameter, so the situation is simple: if you want more stability, reduce `Q`. (or apply `average_fr` with large `F`). This doesn't weaken the theorem in any way.

Scattering doesn't aim for stability in abstracto but with respect to a certain class of transformations. Remember that the constant representation `Sx = 0` satisfies `|| Sx - Sx_tau || <= C ||x||` for any `x`, for any `tau`, and for `C` as small as you wish. Doesn't mean it's a useful representation.

> t - tau as opposed to tau * t

wait, is `tau` a diffeomorphism or a constant real number here? if the latter, you're describing a time shift ... right?

> " the Euclidean distance will grow [...] the only role of 1/Q seems to be to guarantee linearity"

It would help if you put these ideas in equations. There's a limit to how much math i can understand when it's described with English words.

What i can say right now is that your statement:
"The point is that the distance will grow with size of deformation, below or over 1/Q. "
is correct but obvious. Were you expecting that the distance would ever decrease with the size of deformation?
(where "size" here means `K(tau)` )

**Reply by OverLordGoldDragon** [reply]
> only shifts the bounds (takes longer to reach max distance)

Illustrating, `conv(x, phi)` with various supports on `x=|WGN|` (`len(x)==2048`):

![image](https://user-images.githubusercontent.com/16495490/133941978-39c4712c-1320-4346-832e-c9e9fe14dac3.png)

<img src="https://user-images.githubusercontent.com/16495490/133941988-214c071b-04d5-4679-b13d-d0182f92976a.png" width="450">

Black vlines at `T=32, 128` to show region of linearity. `support(x)` means `x` is zero everywhere else. Blue, orange, green, red -> `support(x)` = 128, ..., 2048.

This confirms `T: 32 -> 128` shifts all bounds and "delays" (also reduces depending on kind of deformation) peak distance (`sqrt(2)` since relative distance on inputs of same norm). This will hold _for all_ `x`, so distance bound depends only on: 1) deformation size; 2) filterbank - which is the point of the theorem.

<details>
  <summary><b>code</b></summary>

```python
import numpy as np
from numpy.fft import fft, ifft, ifftshift
from kymatio.visuals import plot, imshow
from kymatio.scattering1d.filter_bank import gauss_1d


def l2(a, b):
    return np.linalg.norm(a - b) / np.linalg.norm(a)

#%%
np.random.seed(0)

N, T = 2048, 64
Ts = np.array([32, 128])
gs = np.array([ifft(gauss_1d(N, .1 / T)).real for T in Ts])
assert np.all(N / Ts >= 16), "poorly decayed lowpass"
x_orig = np.random.randn(N)

x_supps = np.round(np.logspace(np.log10(Ts.max()), np.log10(N), 4)).astype(int)
x_all = []
outs_all = []
for x_width in x_supps:
    x = x_orig.copy()
    x[:N//2-x_width//2] = 0
    x[N//2+x_width//2:] = 0
    x = np.abs(x)
    x /= np.linalg.norm(x)

    outs = ifft(fft(gs) * fft(x)).real
    outs_all.append(outs)
    x_all.append(x)

outs_all, x_all = np.array(outs_all), np.array(x_all)
#%%
mx = np.max(outs_all) * 1.02
for i, T in enumerate(Ts):
    title= "conv(x, phi), support(x)={} | T={}".format(
        ', '.join(map(str, x_supps)), T)
    kw = dict(show=1, title=title, ylims=(-mx/100, mx))
    if i % 2 == 1:
        kw['linestyle'] = '--'
    plot(outs_all[:, i].T, **kw)

#%%
dists_all = []
for supp_idx in range(len(x_supps)):
    dists_all.append([])
    for T_idx, o in enumerate(outs_all[supp_idx]):
        dist = np.array([l2(o, np.roll(o, i)) for i in range(N//2)])
        dists_all[-1].append(dist)

dists_all = np.array(dists_all)
print(dists_all.shape)
#%%
colors = [f'tab:{c}' for c in "blue orange green red purple".split()]
for i, T in enumerate(Ts):
    for j, x_supp in enumerate(x_supps):
        linestyle = '--' if (i % 2 == 1) else '-'
        kw = dict(title="reldists", xlabel="time-shift", linestyle=linestyle,
                  auto_xlims=0, color=colors[j])
        plot(dists_all[j, i], **kw)
        vkw = dict(color='k', linestyle=linestyle, linewidth=2)
        plot([], vlines=(T, vkw))
```
</details>

**Reply by lostanlen** [reply]
and what's your question?

**Reply by OverLordGoldDragon** [reply]
> what makes it unclear?

that `1` can mean `1` or `1/Q`, we can't tell

> Right after the introduction

Again if `1/Q` is silent I have no way of telling. What'd qualify is a step in the theorem that's premised on `1/Q`.

> This doesn't weaken the theorem in any way.

A representation that's stable only for `<1/Q` is much weaker than one that's stable globally. It means for warps of sizes `>1/Q` the distance becomes dependent upon characteristics of `x` besides its norm. My previous post shows that stability holds globally for white noise of various supports - can you show a counterexample?

> is tau a...

I use `tau` in the most general sense - but previous post only shows time-shifts.

> It would help if you put these ideas in equations

More precisely, the _upper bound_ on coefficient distance is determined by size of deformation, where "size" is defined by `K(tau(t))`. So no matter what `x` we pick, its distance with `x_tau` won't exceed what's stated by the theorem. "Linearity" means `distance = K * tau(t)`, where `K` is a proportionality constant - i.e. actual distance, not its upper bound, grows directly with `tau(t)`, within a constant scaling.

> correct but obvious

It's exactly what motivated my digging. "There exists a `C`" was also such a statement. The correct statement is that the _upper bound_ grows with size of deformation; the actual distance can be anything (unless we're in linear zone).

**Reply by OverLordGoldDragon** [reply]
> and what's your question?

Where is it shown that if `1/Q < |tau'| < 1`, the theorem no longer holds?

**Reply by lostanlen** [reply]
> Again if 1/Q is silent I have no way of telling

Again, the theorem assumes `Q=1`. If `Q>1` the theorem does not apply. It's that simple.
If you want to reason with `Q>1`, you need to write another theorem that doesn't assume `Q=1`, then see what assumptions on `tau` you need for the formula you have in mind for `Sx`.

> What'd qualify is a step in the theorem that's premised on 1/Q.

The theorem assumes `Q=1`, so this step does not exist.

> A representation that's stable only for <1/Q is much weaker than one that's stable globally.

The theorem assumes `Q=1`, so it doesn't predict that stability of representations with `Q>1`.
What do you mean by global stability?

> My previous post shows that stability holds globally for white noise of various supports - can you show a counterexample?

I don't understand this because i don't understand global stability. So no i cannot show a counterexample.
But bear in mind that there is no possible comparison between the theorem and your experiments. The theorem provides an upper bound to Euclidean distance, not an approximation. Visualizing a curve isn't enough to tell you if the theoretical bound is tight or not. You'd also have to display the bound. And besides, it's only one upper bound: there might be better ones. There is no proof that this is the best possible upper bound.

What i don't understand is whether you think the theorem is too weak (upper bound is too large) or whether it is false (upper bound is too small and thus violated). But you also seem to make an excessive interpretation of the theorem by setting `Q>1`, which is contrary to what the theorem assumes.

Can you help me clarify what you want?

> More precisely, the upper bound on coefficient distance is determined by size of deformation, where "size" is defined by K(tau(t)). 

yes.

> So no matter what x we pick, its distance with x_tau won't exceed what's stated by the theorem.

yes.


> "Linearity" means distance = K * tau(t), where K is a proportionality constant 

No. The distance between x and x_tau does not depend on t, so what you wrote can't be right.
If you meant `distance = K * sup tau` then again no because `tau(t) = a t + b` has an infinite `sup`, yet the distance between `x` and `x_tau` will be finite.

What linearity means (in our case) is that the set of feature vectors `{ Sx_tau }` for every `tau` will form a (quasi-)linear subspace of the feature space of scattering representations. This can be quantified locally in terms of Laplacian, following what Han Han did in the wav2shape paper. (we were working with tau's that weren't diffeomorphisms so we weren't sure what we'd find)

Useful complement to GIS is this paper by Bietti on stability in CNNs:
---> https://jmlr.org/papers/volume20/18-190/18-190.pdf

This is a very important and profound consequence and (in my opinion) why scattering transform works well together with shallow machine learning.

> It's exactly what motivated my digging.

Your statement was obvious because you stated that distance would grow with size of deformation. 
Mallat's statement is non-obvious because it gives an upper bound on the distance which grows in proportion with size of deformation (under some nontrivial definition `K(tau)` for what size of deformation means). 

>  The correct statement is that the upper bound grows with size of deformation; the actual distance can be anything 

Yes ... anything between zero and the upper bound!

> (unless we're in linear zone).

The theorem doesn't speak of a linear zone. The theorem applies for every `tau` as long as `sup nabla tau <= 1/2`.
What's linear zone for you?
What do you mean by "the distance can't be anything in the linear zone"?

> Where is it shown that if 1/Q < |tau'| < 1, the theorem no longer holds?

It's not shown, but i doubt it would hold with the same assumptions. At the very least, you'd have to rework the `sup nabla tau <= 1/2` assumption and replace it by `sup nabla tau <= 2**(-1/Q)`.
I imagine you'd have to look at the proof of Lemma 2.11 and make the necessary changes. It's far from trivial.

Eventually i am convinced that a comparable theorem to Theorem 2.12 could be proven for `Q>1` but that hasn't been formulated nor proven yet. What's important to recognize is that this `Q>1` theorem cannot simply be derived as a consequence from Theorem 2.12 but would need a rewritten proof (including for some of its lemmata like 2.11).

**Reply by OverLordGoldDragon** [reply]
> The theorem assumes Q=1, so this step does not exist.

If no step in the theorem is invalidated by `Q != 1`, then the theorem either isn't proven, or it remains true.


> At the very least, you'd have to rework the ...

By that logic we also can't know when, if ever, the theorem holds for `Q != 1`? If it does hold, why for precisely `1/Q`? Or if we were to rework it, why by assuming `< 1/Q` for `|tau'|`?

> Lemma 2.11

is for lowpass filtering, not related to `Q`, so I see no connection?

> The distance between x and x_tau does not depend on t

I meant pointwise distance, and linearity in sense of linear distance - and should've used `K(tau(t))` instead, but looking back, I actually meant the exact formulation in theorem except replacing inequality by equality. But admittedly I wasn't sure what was meant by "linearizing".

> (quasi-)linear subspace ... Useful complement

I'll take a look, thanks

> Mallat's statement is non-obvious

It's obvious if we just state "there exists a constant", which is how the theorem and papers citing it formulate, as we can always pick `C` such that `<= ||x|| + ||x_tau||`; what's non-obvious is that `C` is determined entirely by filterbank.

> What i don't understand is whether you think the theorem is too weak (upper bound is too large) or whether it is false (upper bound is too small and thus violated)

I think it is true and holds for all `|tau'| < 1`. It is "too weak" if it falls apart for `> 1/Q` - i.e., we know of no bounds on distance except `||x|| + ||x_tau||`.

> I don't understand this because i don't understand global stability. So no i cannot show a counterexample.

By "global" I mean holding for all `|tau'| < 1`. Thus it suffices to find any finite example, with any `Q != 1` and `tau`, where distance grows faster than predicted by the theorem.

> there is no possible comparison between the theorem and your experiments

A transform designed for finite applications cannot be verified by finite examples? Then I don't see how the theorem's useful.

DSS writes, _"the constant `C` is inversely proportional to octave bandwidth of wavelet filters"_ - i.e. `C ~ Q`, which makes sense; then why additionally require `|tau'| < 1/Q`? Your citation reads, _"with no loss of generality"_ - and DSS mentions dependence on `Q` but not of `|tau'|`, yet there is? That's false advertising.

But I don't think so. If there's indeed no proof for `|tau'| > 1/Q`, then I'll gladly make a conjecture that cites _"was never falsified"_.

Further, I think I found such steps - the very first one in Lemma 2.14:

<img src="https://user-images.githubusercontent.com/16495490/134120539-f010fc80-13a3-43e8-aea0-61d9558e8322.png" width="580">

Why `2^j`? Nothing about wavelets has `2` built-in - so the general form must be `2^(j/Q)`, and then we'll find a step that's falsified with `Q != 1`. Depending on the step, it may or may not show that `|tau'| < 1` doesn't hold, or that it must be precisely `< 1/Q`.

**Reply by OverLordGoldDragon** [reply]
I figure the theoretical `phi` does depend on `Q` else it'd break LP sum and tight frames, and it omits in implementation since lower freqs are tiled per STFT - and there is a `2^J` term, with a non-hardwired `2`.
